File: solutions/haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

import Data.Set as Set

multiples :: Integer -> Integer -> [Integer]
multiples 0 _ = [0]
multiples n limit = takeWhile (< limit) [n, 2*n ..]

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples bases limit =
    let allMultiples = concatMap (`multiples` limit) bases
        unique       = Set.fromList allMultiples
    in sum unique

File: solutions/csharp/sum-of-multiples/SumOfMultiples.cs
public static class SumOfMultiples
{
    public static int Sum(IEnumerable<int> multiples, int max)
    {
        List<int> multiples_list = [];
        foreach (int base_value in multiples)
        {
            if (base_value == 0)
            {
                multiples_list.Add(0);
                continue;
            }
            int multiple = base_value;
            while(multiple < max)
            {
                multiples_list.Add(multiple);
                multiple += base_value;
            }
        }
        return multiples_list.Distinct().Sum();
    }
}